### ISSUE TITLE
Kill ffmpeg/ffprobe processes if parent .NET application exits

### ISSUE THREAD

**Issue by lostmsu** [rosenbjerg/FFMpegCore#333]
More generally, if .NET process that launched ffmpeg exits, ffmpeg does not, and there's no option to make it. Most annoying when stopping the .NET app from debugger, as the app has no chance to do anything.

Now I considered contributing this, but
- would like to consult with maintainer first on how to approach it (esp. with Instances project)
- too lazy to port this to Unix-like OSes

On Windows in a similar situation this did the trick: https://gist.github.com/lostmsu/a806fbfedf7a3dba184c4e3d38dc5110

**Comment by hqkirkland** [comment]
I’m trying to do something similar— I’m building a .NET Core Worker that reads a streaming schedule, and uses this (AWESOME!!) wrapper lib to handle all the commands.

I tried to use the `stoppingToken` provided by `ExecuteAsync` as the option for `CancellableThrough`, but stopping my debugging doesn’t stop the ffmpeg process. This is a pain because I’m an (overly) aggressive debugger & I keep needing to kill it manually.

My guess was that VS is just ending the main process and not truly “stopping” the service, thus ffmpeg continues.. but I’m not sure. Not currently at home to investigate.

Btw side note: is Console output for the child process supported? I’d like to use the built-in Logger for worker services to monitor the process.

**Comment by pedrosbanioles** [comment]
If you are using .NET Framework, you could perhaps hook into the AppDomain.ProcessExit, you would of course need to try to include a reference to your CancellationTokenSource in order to send the cancellation. (https://stackoverflow.com/a/1119869/2573213 may help)

If using .NET core, you could hook into the IApplicationLifetime at startup and add a handler for the OnApplicationStopping event (see here for example: https://stackoverflow.com/a/74956744/2573213). Same principal of accessing the CancellationTokenSource to send the cancellation would apply

**Comment by Loongle** [comment]
any update?

**Comment by NPatch** [comment]
> If you are using .NET Framework, you could perhaps hook into the AppDomain.ProcessExit, you would of course need to try to include a reference to your CancellationTokenSource in order to send the cancellation. (https://stackoverflow.com/a/1119869/2573213 may help)
> 
> If using .NET core, you could hook into the IApplicationLifetime at startup and add a handler for the OnApplicationStopping event (see here for example: https://stackoverflow.com/a/74956744/2573213). Same principal of accessing the CancellationTokenSource to send the cancellation would apply

>My guess was that VS is just ending the main process and not truly “stopping” the service, thus ffmpeg continues.. but I’m not sure. 

Exactly. VS kills the main process and doesn't bother with children. If we're talking about running the app outside of the IDE, then yes, the aforementioned APIs work. You could also write the children procs to poll for parent termination and terminate if the parent terminates, but in this case, this lib can't do that as it would be part of the terminated main process and the children are ffmpeg instances which do not seem to poll parent state at all (at least nothing I found in the documentation).

One possible solution is for the debugger itself to do that. Just last week the .net debugger in vscode added a new [launch option](https://github.com/dotnet/vscode-csharp/pull/9218) to terminate child processes as well. But VS doesn't seem to support that yet.
